tests/test_asyncio.py

```diff
@@ -30,34 +31,125 @@ async def test_asyncio_bad_put_raises(dummy_server_async):
+        assert isinstance(await data_generator.__anext__(), StartData)
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert not flush_event.is_set()
+        assert isinstance(await wait_and_press_button(data_generator), FrameData)
+        assert isinstance(await data_generator.__anext__(), EndData)
+        await data_generator.aclose()
+
+
+async def test_asyncio_data_flush_every_frame(
+    dummy_server_async, fast_dump, fast_dump_expected
+):
     dummy_server_async.data = fast_dump
+
     events = []
     async with AsyncioClient("localhost") as client:
-        async for data in client.data(frame_timeout=1, flush_period=flush_period):
+        async for data in client.data(
+            frame_timeout=5,
+            flush_mode=FlushMode.IMMEDIATE,
+        ):
             events.append(data)
+
             if len(events) == len(fast_dump_expected):
                 break
+
     assert fast_dump_expected == events
 
 
+async def test_asyncio_data_timeout_on_no_manual_press(dummy_server_async, fast_dump):
+    flush_event = asyncio.Event()
+    async with AsyncioClient("localhost") as client:
+        with pytest.raises(asyncio.TimeoutError, match="No data received for 3s"):
+            async for data in client.data(
+                frame_timeout=3, flush_mode=FlushMode.MANUAL, flush_event=flush_event
+            ):
+                pass
+
+
 async def test_asyncio_data_timeout(dummy_server_async, fast_dump):
     dummy_server_async.data = fast_dump
     async with AsyncioClient("localhost") as client:
         with pytest.raises(asyncio.TimeoutError, match="No data received for 0.1s"):
-            async for data in client.data(frame_timeout=0.1):
+            async for data in client.data(
+                frame_timeout=0.1, flush_mode=FlushMode.IMMEDIATE
+            ):
                 "This goes forever, when it runs out of data we will get our timeout"
 
 
+async def test_asyncio_data_nonexistent_flushmode(dummy_server_async, fast_dump):
+    dummy_server_async.data = fast_dump
+    async with AsyncioClient("localhost") as client:
+        with pytest.raises(ValueError, match="flush_mode must be one of"):
+            async for data in client.data(frame_timeout=0.1, flush_mode=None):
+                pass
+
+
+@pytest.mark.parametrize(
+    "kwargs",
+    [
+        {"flush_mode": FlushMode.PERIODIC, "flush_period": 0},
+        {"flush_mode": FlushMode.MANUAL, "flush_event": None},
+    ],
+)
+async def test_asyncio_data_missing_kwargs(dummy_server_async, fast_dump, kwargs):
+    print(kwargs)
+    dummy_server_async.data = fast_dump
+    async with AsyncioClient("localhost") as client:
+        with pytest.raises(ValueError):
```

Comment:
It's paranoid and pedantic, but is it worth checking that the exception that gets raised is the expected one with the expected message? Otherwise we risk catching `ValueError` raised for unrelated reasons

Reply:
Agreed, I'll implement now